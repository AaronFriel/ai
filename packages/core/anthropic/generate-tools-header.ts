import {
  LanguageModelV1FunctionTool,
  UnsupportedJSONSchemaError,
} from '../spec';

// see https://docs.anthropic.com/claude/docs/functions-external-tools#function-calling-format
export const TOOLS_HEADER_START = `\
In this environment you have access to a set of tools you can use to answer the user's question.

You may call them like this:
<function_calls>
<invoke>
<tool_name>$TOOL_NAME</tool_name>
<parameters>
<$PARAMETER_NAME>$PARAMETER_VALUE</$PARAMETER_NAME>
...
</parameters>
</invoke>
</function_calls>

Here are the tools available:
`;

export function generateToolsHeader({
  tools,
  provider,
}: {
  tools: LanguageModelV1FunctionTool[];
  provider: string;
}): string {
  return (
    TOOLS_HEADER_START +
    tools.map(tool => generateToolXml({ tool, provider })).join('\n\n')
  );
}

function generateToolXml({
  tool,
  provider,
}: {
  tool: LanguageModelV1FunctionTool;
  provider: string;
}): string {
  const schema = tool.parameters;

  if (schema.type !== 'object') {
    throw new UnsupportedJSONSchemaError({
      provider,
      schema,
      reason: 'Tool parameters must be objects.',
    });
  }

  if (schema.properties == null) {
    throw new UnsupportedJSONSchemaError({
      provider,
      schema,
      reason: 'Tool parameter objects must have properties.',
    });
  }

  let text = '<tool_description>\n';
  text += `<tool_name>${tool.name}</tool_name>\n`;

  if (tool.description != null) {
    text += `<description>${tool.description}</description>\n`;
  }

  text += '<parameters>\n';

  const properties = Object.entries(schema.properties);
  for (const [name, property] of properties) {
    text += `<parameter>\n`;
    text += `<name>${name}</name>\n`;

    if (typeof property === 'boolean') {
      throw new UnsupportedJSONSchemaError({
        provider,
        schema,
        reason: 'Tool parameter properties must be objects.',
      });
    }

    text += `<type>${property.type}</type>\n`;
    text += `</parameter>\n`;
  }

  text += '</parameters>\n';
  text += '</tool_description>';

  return text;
}
