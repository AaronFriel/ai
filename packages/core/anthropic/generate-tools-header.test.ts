import { UnsupportedJSONSchemaError } from '../spec';
import { generateToolsHeader } from './generate-tools-header';

it('should throw unsupported schema error when parameter is string', async () => {
  await expect(async () =>
    generateToolsHeader({
      provider: 'anthropic.messages',
      tools: [
        {
          type: 'function',
          name: 'my-function',
          parameters: {
            type: 'string',
            $schema: 'http://json-schema.org/draft-07/schema#',
          },
        },
      ],
    }),
  ).rejects.toThrowError(UnsupportedJSONSchemaError);
});

it('should generate tools header for a single tool with single string parameter', async () => {
  const toolsHeader = generateToolsHeader({
    provider: 'anthropic.messages',
    tools: [
      {
        type: 'function',
        name: 'my-function',
        description: 'my function description.',
        parameters: {
          type: 'object',
          properties: { param1: { type: 'string' } },
          required: ['param1'],
          additionalProperties: false,
          $schema: 'http://json-schema.org/draft-07/schema#',
        },
      },
    ],
  });

  expect(toolsHeader).toMatchSnapshot();
});

it('should generate tools header for two tools with single parameter each', async () => {
  const toolsHeader = generateToolsHeader({
    provider: 'anthropic.messages',
    tools: [
      {
        type: 'function',
        name: 'my-function-1',
        parameters: {
          type: 'object',
          properties: { first: { type: 'boolean' } },
          required: ['first'],
          additionalProperties: false,
          $schema: 'http://json-schema.org/draft-07/schema#',
        },
      },
      {
        type: 'function',
        name: 'my-function-2',
        parameters: {
          type: 'object',
          properties: { second: { type: 'number' } },
          required: ['second'],
          additionalProperties: false,
          $schema: 'http://json-schema.org/draft-07/schema#',
        },
      },
    ],
  });

  expect(toolsHeader).toMatchSnapshot();
});
