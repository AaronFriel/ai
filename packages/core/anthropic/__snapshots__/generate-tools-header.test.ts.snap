// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should generate tools header for a single tool with single string parameter 1`] = `
"In this environment you have access to a set of tools you can use to answer the user's question.

You may call them like this:
<function_calls>
<invoke>
<tool_name>$TOOL_NAME</tool_name>
<parameters>
<$PARAMETER_NAME>$PARAMETER_VALUE</$PARAMETER_NAME>
...
</parameters>
</invoke>
</function_calls>

Here are the tools available:
<tool_description>
<tool_name>my-function</tool_name>
<description>my function description.</description>
<parameters>
<parameter>
<name>param1</name>
<type>string</type>
</parameter>
</parameters>
</tool_description>"
`;

exports[`should generate tools header for two tools with single parameter each 1`] = `
"In this environment you have access to a set of tools you can use to answer the user's question.

You may call them like this:
<function_calls>
<invoke>
<tool_name>$TOOL_NAME</tool_name>
<parameters>
<$PARAMETER_NAME>$PARAMETER_VALUE</$PARAMETER_NAME>
...
</parameters>
</invoke>
</function_calls>

Here are the tools available:
<tool_description>
<tool_name>my-function-1</tool_name>
<parameters>
<parameter>
<name>first</name>
<type>boolean</type>
</parameter>
</parameters>
</tool_description>

<tool_description>
<tool_name>my-function-2</tool_name>
<parameters>
<parameter>
<name>second</name>
<type>number</type>
</parameter>
</parameters>
</tool_description>"
`;
